### ISSUE TITLE
About the lengths of train_loader and unlabeled_loader

### ISSUE THREAD

**Issue by chengjianhong** [FlyreelAI/sesemi#4]
Hi, are the lengths of train_loader and unlabeled_loader the same in your research? What if they are not equal?

**Comment by vuptran** [comment]
The length of `unlabeled_loader` will almost always be larger than `train_loader` because in practical applications, there will almost always be more unlabeled than labeled data. As such, we iterate through the `unlabeled_loader` and catch the `StopIteration` exception in the `train()` function (https://github.com/FlyreelAI/sesemi/blob/master/utils.py#L63). The way we set it up currently is to have unlabeled data consisting of all train, val, and optional `--unlabeled-dir`. So one training epoch means one pass over the `unlabeled_loader`, which could mean many passes over the `train_loader`.

**Comment by chengjianhong** [comment]
OK. I can understand the unlabeled_loader? what is the class label?  It should be rotation labels but it can't find the relationship in unlabeled_loader. Moreover, In train_loader , I can't understand the "args.batch_size * unlabeled_classes".
```
unlabeled_loader = torch.utils.data.DataLoader(
        unlabeled_dataset,
        batch_size=args.batch_size, shuffle=True,
        num_workers=args.workers, pin_memory=True,
        worker_init_fn=lambda x: np.random.seed(),
        collate_fn=rotate, drop_last=True)
```

```
rotate = RotationTransformer()
    unlabeled_classes = rotate.rotation_labels
```
```
train_loader = torch.utils.data.DataLoader(
        train_dataset,
        batch_size=args.batch_size * unlabeled_classes,
        shuffle=True, num_workers=args.workers, pin_memory=True,
        worker_init_fn=lambda x: np.random.seed(), drop_last=True)
```

**Comment by vuptran** [comment]
The training class labels are taken from the folder names using the PyTorch ImageFolder data loader. The unlabeled class labels are 4 rotations derived from the `unlabeled_dataset` with `collate_fn=rotate` (https://github.com/FlyreelAI/sesemi/blob/master/open_sesemi.py#L127). Here, `rotate.rotation_labels = 4`. In `train_loader`, the batch size is set to be 4x the argument batch size, which is not required because you can set the batch size to be an arbitrary number in `train_loader` as GPU memory allows.

**Comment by chengjianhong** [comment]
OK, Thank you very much.
Recently I am studying semi-supervised learning. If unlabeled data gets a pseudo-label with High confidence from a pre-trained model, I want to add the (unlabeled data, pseudo-label) into the train_loader for further training the model. How to add these data into the train_loader and load it efficiently? Can you help me and give me an example?

**Comment by vuptran** [comment]
First pull the latest from this repo to enable the functionality.

1. Train your model using the latest code
2. Modify and adapt the `inference.py` code to your use case for predicting pseudo labels
3. You can choose to create a new folder to store your pseudo labels, something like this structure: `/path/to/pseudo-labels/train`
4. Each subdirectory in your pseudo folder is your predicted pseudo class
5. Re-train your model by passing a list of the original labeled directory and the pseudo directory into the argument `--data-dir`, which now accepts a list of inputs. Something like this: `python -u fully_supervised.py --data-dir /path/to/labeled/dir /path/to/pseudo-labels`
6. Make sure your pseudo directory has the same structure as the original labeled directory, or the code will raise an error of mismatched class names. If your pseudo directory has fewer class names than the labeled directory, you can create empty folders for it. The code will ignore empty folders for training, but need to have the same structure as the original training set.
7. Happy training!

**Comment by vuptran** [comment]
Closing this as it is not really an issue. Feel free to open a [discussion](https://github.com/FlyreelAI/sesemi/discussions) about  it, along with others, as you use the codebase.
